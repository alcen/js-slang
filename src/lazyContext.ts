import { Context } from './types'

export const LAZY_SOURCE_1 = 2001
export const LAZY_SOURCE_2 = 2002

/**
 * Checks whether the interpreter should evaluate
 * the expression lazily in this context. Currently,
 * lazy evaluation is only implemented for Source
 * chapters 1 and 2.
 *
 * @param context The context to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export default function lazyEvaluate(context: Context): boolean {
  return lazyEvaluateInChapter(context.chapter)
}

/**
 * Checks if in the current context, the program will be
 * transpiled with the lazy evaluation transpiler.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateInTranspiler(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'native'
}

/**
 * Checks if in the current context, the program will be
 * interpreted with the lazy evaluation interpreter.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateInInterpreter(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'interpreter'
}

/**
 * Checks if we have to dynamically determine which execution
 * method to use in this context for lazy evaluation.
 *
 * @param context The context to be checked.
 */
export function lazyEvaluateAuto(context: Context): boolean {
  return lazyEvaluate(context) && context.executionMethod === 'auto'
}

/**
 * Checks whether the interpreter should evaluate the
 * expression lazily, given a chapter number.
 *
 * @param chapter The chapter to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export function lazyEvaluateInChapter(chapter: number): boolean {
  return lazyEvaluateInSource1(chapter) || lazyEvaluateInSource2(chapter)
}

/**
 * Checks whether this chapter number represents
 * Lazy Source chapter 1 (functions, arithmetic,
 * conditionals).
 *
 * @param chapter The chapter number.
 */
export function lazyEvaluateInSource1(chapter: number): boolean {
  return chapter === LAZY_SOURCE_1
}

/**
 * Checks whether this chapter number represents
 * Lazy Source chapter 2 (pairs and lists).
 *
 * @param chapter The chapter number.
 */
export function lazyEvaluateInSource2(chapter: number): boolean {
  return chapter === LAZY_SOURCE_2
}
