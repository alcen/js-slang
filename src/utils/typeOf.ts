import { Value } from "../types"

// We need to define our own typeof in order for null/array to display properly in error messages
export const typeOf = (v: Value) => {
  if (v === null) {
    return 'null'
  } else if (Array.isArray(v)) {
    return 'array'
  } else {
    return typeof v
  }
}
