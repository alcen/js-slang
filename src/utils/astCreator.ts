import * as es from 'estree'
import { astThunkNativeTag } from '../stdlib/lazy'
import { AllowedDeclarations, BlockExpression, FunctionDeclarationExpression } from '../types'
import { typeOf } from './typeOf'

export const locationDummyNode = (line: number, column: number) =>
  literal('Dummy', { start: { line, column }, end: { line, column } })

export const identifier = (name: string): es.Identifier => ({
  type: 'Identifier',
  name
})

/**
 * Constructs a new Thunk value for lazy evaluation. This function
 * does this by taking the original Literal in the Abstract Syntax
 * Tree, and changing it into an ArrowFunctionExpression with the
 * literal in its body
 * @param node The Literal to be transformed.
 */
export const mutateToThunk = (node: es.Literal): es.ObjectExpression => {
  // creates a new copy of node.loc
  function copyLoc() {
    if (!node.loc) {
      return undefined
    } else {
      return {
        start: Object.assign({}, node.loc.start),
        end: Object.assign({}, node.loc.end)
      }
    }
  }
  // make a new object for the new Thunk
  const newNode = node as any
  newNode.type = 'ObjectExpression'
  // get type of literal
  const type = typeOf(node.value)
  // ensure that the raw type becomes a string, not a variable
  const typeRaw = '"' + type + '"'
  // get rid of old value and old raw
  const oldValue = newNode.value
  newNode.value = undefined
  const oldRaw = newNode.raw
  newNode.raw = undefined
  newNode.properties = []
  // set Thunk properties
  // first, create type property
  newNode.properties[0] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'type'
    },
    // value of type is a string literal
    value: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: type,
      raw: typeRaw
    },
    kind: 'init'
  }
  // get arrow function
  const arrowFunction = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: oldValue,
      raw: oldRaw
    }
  }
  // then, create the value property
  // also create lambda function storing original literal
  newNode.properties[1] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'value'
    },
    // value of 'value' is an ArrowFunctionExpression
    value: arrowFunction,
    kind: 'init'
  }
  // get toString to show the normal result
  const toString = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: JSON.stringify(oldValue),
      raw: JSON.stringify(oldRaw)
    },
    // add tag to prevent toString() from getting wrapped in
    // wrapArrowFunctionsToAllowNormalCallsAndNiceToString
    // (toString is already giving it a nice string representation)
    tag: astThunkNativeTag
  }
  // lastly create the toString property
  // so thunks appear as normal values
  newNode.properties[2] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'toString'
    },
    value: toString,
    kind: 'init'
  }
  return newNode
}

export const literal = (value: string | number | boolean, loc?: es.SourceLocation): es.Literal => ({
  type: 'Literal',
  value,
  loc
})

export const memberExpression = (
  object: es.Expression,
  propertyString: string
): es.MemberExpression => ({
  type: 'MemberExpression',
  object,
  computed: false,
  property: identifier(propertyString)
})

export const declaration = (
  name: string,
  kind: AllowedDeclarations,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  declarations: [
    {
      type: 'VariableDeclarator',
      id: identifier(name),
      init
    }
  ],
  kind,
  loc
})

export const constantDeclaration = (name: string, init: es.Expression, loc?: es.SourceLocation) =>
  declaration(name, 'const', init, loc)

export const callExpression = (
  callee: es.Expression,
  args: es.Expression[],
  loc?: es.SourceLocation
): es.CallExpression => ({
  type: 'CallExpression',
  callee,
  arguments: args,
  loc
})

export const expressionStatement = (expression: es.Expression): es.ExpressionStatement => ({
  type: 'ExpressionStatement',
  expression
})

export const blockArrowFunction = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const functionExpression = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionExpression => ({
  type: 'FunctionExpression',
  id: null,
  async: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const blockStatement = (body: es.Statement[]): es.BlockStatement => ({
  type: 'BlockStatement',
  body
})

export const program = (body: es.Statement[]): es.Program => ({
  type: 'Program',
  sourceType: 'module',
  body
})

export const returnStatement = (
  argument: es.Expression,
  loc?: es.SourceLocation
): es.ReturnStatement => ({
  type: 'ReturnStatement',
  argument,
  loc
})

export const property = (key: string, value: es.Expression): es.Property => ({
  type: 'Property',
  method: false,
  shorthand: false,
  computed: false,
  key: identifier(key),
  value,
  kind: 'init'
})

export const objectExpression = (properties: es.Property[]): es.ObjectExpression => ({
  type: 'ObjectExpression',
  properties
})

export const mutateToCallExpression = (
  node: es.Node,
  callee: es.Expression,
  args: es.Expression[]
) => {
  node.type = 'CallExpression'
  node = node as es.CallExpression
  node.callee = callee
  node.arguments = args
}

export const logicalExpression = (
  operator: es.LogicalOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.LogicalExpression => ({
  type: 'LogicalExpression',
  operator,
  left,
  right,
  loc
})

export const conditionalExpression = (
  test: es.Expression,
  consequent: es.Expression,
  alternate: es.Expression,
  loc?: es.SourceLocation
): es.ConditionalExpression => ({
  type: 'ConditionalExpression',
  test,
  consequent,
  alternate,
  loc
})

export const arrayExpression = (elements: es.Expression[]): es.ArrayExpression => ({
  type: 'ArrayExpression',
  elements
})

export const assignmentExpression = (
  left: es.Identifier | es.MemberExpression,
  right: es.Expression
): es.AssignmentExpression => ({
  type: 'AssignmentExpression',
  operator: '=',
  left,
  right
})

export const binaryExpression = (
  operator: es.BinaryOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.BinaryExpression => ({
  type: 'BinaryExpression',
  operator,
  left,
  right,
  loc
})

export const unaryExpression = (
  operator: es.UnaryOperator,
  argument: es.Expression,
  loc?: es.SourceLocation
): es.UnaryExpression => ({
  type: 'UnaryExpression',
  operator,
  prefix: true,
  argument,
  loc
})

// primitive: undefined is a possible value
export const primitive = (value: any): es.Expression => {
  return value === undefined ? identifier('undefined') : literal(value)
}

export const functionDeclarationExpression = (
  id: es.Identifier,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): FunctionDeclarationExpression => ({
  type: 'FunctionExpression',
  id,
  params,
  body,
  loc
})

export const functionDeclaration = (
  id: es.Identifier | null,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionDeclaration => ({
  type: 'FunctionDeclaration',
  id,
  params,
  body,
  loc
})

export const blockExpression = (
  body: es.Statement[],
  loc?: es.SourceLocation
): BlockExpression => ({
  type: 'BlockExpression',
  body,
  loc
})

export const arrowFunctionExpression = (
  params: es.Pattern[],
  body: es.Expression | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: body.type !== 'BlockStatement',
  generator: false,
  params,
  body,
  loc
})

export const variableDeclaration = (
  declarations: es.VariableDeclarator[],
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  kind: 'const',
  declarations,
  loc
})

export const variableDeclarator = (
  id: es.Pattern,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclarator => ({
  type: 'VariableDeclarator',
  id,
  init,
  loc
})

export const ifStatement = (
  test: es.Expression,
  consequent: es.BlockStatement,
  alternate: es.Statement,
  loc?: es.SourceLocation
): es.IfStatement => ({
  type: 'IfStatement',
  test,
  consequent,
  alternate,
  loc
})
