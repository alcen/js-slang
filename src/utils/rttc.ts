import * as es from 'estree'
import { RuntimeSourceError } from '../interpreter-errors'
import { Thunk } from '../stdlib/lazy'
import { ErrorSeverity, ErrorType, Value } from '../types'
import { typeOf } from './typeOf'

const LHS = ' on left hand side of operation'
const RHS = ' on right hand side of operation'

export class TypeError extends RuntimeSourceError {
  public type = ErrorType.RUNTIME
  public severity = ErrorSeverity.ERROR
  public location: es.SourceLocation

  constructor(node: es.Node, public side: string, public expected: string, public got: string) {
    super(node)
  }

  public explain() {
    return `Expected ${this.expected}${this.side}, got ${this.got}.`
  }

  public elaborate() {
    return this.explain()
  }
}

// Checks if a Thunk is a number
const isNumberT = (v: Thunk<Value>) => v.type === 'number'
// Checks if Thunk is a string
const isStringT = (v: Thunk<Value>) => v.type === 'string'

const isNumber = (v: Value) => typeOf(v) === 'number'
// See section 4 of https://2ality.com/2012/12/arrays.html
// v >>> 0 === v checks that v is a valid unsigned 32-bit int
// tslint:disable-next-line:no-bitwise
const isArrayIndex = (v: Value) => isNumber(v) && v >>> 0 === v && v < 2 ** 32 - 1
const isString = (v: Value) => typeOf(v) === 'string'
const isBool = (v: Value) => typeOf(v) === 'boolean'
const isObject = (v: Value) => typeOf(v) === 'object'
const isArray = (v: Value) => typeOf(v) === 'array'

export const checkUnaryExpression = (node: es.Node, operator: es.UnaryOperator, value: Value) => {
  if ((operator === '+' || operator === '-') && !isNumber(value)) {
    return new TypeError(node, '', 'number', typeOf(value))
  } else if (operator === '!' && !isBool(value)) {
    return new TypeError(node, '', 'boolean', typeOf(value))
  } else {
    return undefined
  }
}

export const checkBinaryExpression = (
  node: es.Node,
  operator: es.BinaryOperator,
  left: Thunk<Value>,
  right: Thunk<Value>
) => {
  switch (operator) {
    case '-':
    case '*':
    case '/':
    case '%':
      if (!isNumberT(left)) {
        return new TypeError(node, LHS, 'number', left.type)
      } else if (!isNumberT(right)) {
        return new TypeError(node, RHS, 'number', right.type)
      } else {
        return 'number'
      }
    case '+':
      if (isNumberT(left)) {
        if (isNumberT(right)) {
          return 'number'
        } else {
          return new TypeError(node, RHS, 'number', right.type)
        }
      } else if (isStringT(left)) {
        if (isStringT(right)) {
          return 'string'
        } else {
          return new TypeError(node, RHS, 'string', right.type)
        }
      } else {
        return new TypeError(node, LHS, 'string or number', left.type)
      }
    case '<':
    case '<=':
    case '>':
    case '>=':
      if (isNumberT(left)) {
        if (isNumberT(right)) {
          return 'boolean'
        } else {
          return new TypeError(node, RHS, 'number', right.type)
        }
      } else if (isStringT(left)) {
        if (isStringT(right)) {
          return 'boolean'
        } else {
          return new TypeError(node, RHS, 'string', right.type)
        }
      } else {
        return new TypeError(node, LHS, 'string or number', left.type)
      }
    case '!==':
    case '===':
      return 'boolean'
    default:
      // unknown type, may still be valid due to conditionals
      return ''
  }
}

export const checkIfStatement = (node: es.Node, test: Value) => {
  return isBool(test) ? undefined : new TypeError(node, ' as condition', 'boolean', typeOf(test))
}

export const checkMemberAccess = (node: es.Node, obj: Value, prop: Value) => {
  if (isObject(obj)) {
    return isString(prop) ? undefined : new TypeError(node, ' as prop', 'string', typeOf(prop))
  } else if (isArray(obj)) {
    return isArrayIndex(prop)
      ? undefined
      : isNumber(prop)
      ? new TypeError(node, ' as prop', 'array index', 'other number')
      : new TypeError(node, ' as prop', 'array index', typeOf(prop))
  } else {
    return new TypeError(node, '', 'object or array', typeOf(obj))
  }
}
