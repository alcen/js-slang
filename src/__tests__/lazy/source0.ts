import { stripIndent } from '../../utils/formatters'
import { expectResult } from '../../utils/testing'
import { LAZY_SOURCE_1 } from '../../lazyContext'
import { ExecutionMethod } from '../../types'

// ================= UTILITY FUNCTIONS =================
const runTestSuccess = (
  executionMethod: ExecutionMethod,
  program: string,
  result: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectResult(stripIndent(program), {
      chapter: LAZY_SOURCE_1,
      executionMethod
    }).toMatchInlineSnapshot(result)
  })

// ==================== LAZY TESTS =====================
const additionTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
    43694 + 89076;
  `,
    '132770',
    'addition'
  )

const subtractionTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
    83553 - 28018;
  `,
    '55535',
    'subtraction'
  )

const multiplicationTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  95533 * 269;
`,
    '25698377',
    'multiplication'
  )

const divisionTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  69573 / 7;
`,
    '9939',
    'division'
  )

const equalityTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  4 === 423;
`,
    'false',
    'equality'
  )

const inequalityTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  577 !== 577;
`,
    'false',
    'inequality'
  )

const logicalAndTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  true && false;
`,
    'false',
    'logical and'
  )

const logicalOrTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  true || error();
`,
    'true',
    'logical Or'
  )

const logicalAndTestLazy = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  false && error();
`,
    'false',
    'logical and (lazy)'
  )

const conditionalsTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  false ? 20 : 46;
`,
    '46',
    'conditionals test'
  )

const conditionalsTestLazyConsequent = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  false ? error() : 934;
`,
    '934',
    'conditionals test (lazy consequent)'
  )

const conditionalsTestLazyAlternative = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  true ? 782 : error();
`,
    '782',
    'conditionals test (lazy alternative)'
  )

const nestedExpressionsTest = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
  (true && (true || false)) ? (441 * 2) : (7 + 245);
`,
    '882',
    'nested expressions test'
  )

// ================== RUN THE TESTS ====================
const testArray = [
  additionTest,
  subtractionTest,
  multiplicationTest,
  divisionTest,
  equalityTest,
  inequalityTest,
  logicalAndTest,
  logicalOrTest,
  logicalAndTestLazy,
  conditionalsTest,
  conditionalsTestLazyConsequent,
  conditionalsTestLazyAlternative,
  nestedExpressionsTest
]
testArray.forEach(tst => tst('interpreter'))
testArray.forEach(tst => tst('native'))
testArray.forEach(tst => tst('auto'))
