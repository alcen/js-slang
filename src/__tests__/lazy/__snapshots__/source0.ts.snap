// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addition (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "43694 + 89076;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 132770,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`addition (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "43694 + 89076;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 132770,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`addition (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "43694 + 89076;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 132770,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? 20 : 46;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 46,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? 20 : 46;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 46,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy alternative) (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true ? 782 : error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 782,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy alternative) (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true ? 782 : error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 782,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy alternative) (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true ? 782 : error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 782,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy consequent) (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? error() : 934;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 934,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy consequent) (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? error() : 934;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 934,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (lazy consequent) (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? error() : 934;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 934,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`conditionals test (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false ? 20 : 46;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 46,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`division (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "69573 / 7;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 9939,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`division (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "69573 / 7;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 9939,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`division (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "69573 / 7;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 9939,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`equality (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "4 === 423;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`equality (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "4 === 423;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`equality (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "4 === 423;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`inequality (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "577 !== 577;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`inequality (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "577 !== 577;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`inequality (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "577 !== 577;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical Or (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true || error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical Or (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true || error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical Or (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true || error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true && false;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true && false;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (lazy) (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false && error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (lazy) (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false && error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (lazy) (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "false && error();",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`logical and (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true && false;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`multiplication (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "95533 * 269;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 25698377,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`multiplication (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "95533 * 269;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 25698377,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`multiplication (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "95533 * 269;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 25698377,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`nested expressions test (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "(true && (true || false)) ? (441 * 2) : (7 + 245);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 882,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`nested expressions test (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "(true && (true || false)) ? (441 * 2) : (7 + 245);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 882,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`nested expressions test (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "(true && (true || false)) ? (441 * 2) : (7 + 245);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 882,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`subtraction (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "83553 - 28018;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 55535,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`subtraction (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "83553 - 28018;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 55535,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`subtraction (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "83553 - 28018;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 55535,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
